src/helpers/circular_map.rs

```diff
@@ -82,7 +72,11 @@ impl<K: Clone + PartialEq, V: Clone, const N: u32> CircularMap<K, V, N> {
     ///
     pub fn insert(&mut self, key: K, value: V) -> bool {
         if !self.contains_key(&key) {
-            self.queue.push((key, value));
+            if let Some(Some(popped)) = self.queue.push(Some(key.clone())) {
+                self.map.remove(&popped);
+            }
+            self.map.insert(key, value);
```

Comment:
performance note: since we are using this collection only for unconfirmed blocks (where the key is a block hash and the value is the corresponding block), this method could be made faster by doing the following:
```
if self.map.insert(key, value).is_none() {
    if let Some(Some(popped)) = self.queue.push(Some(key.clone())) {
        self.map.remove(&popped);
    }
    // insertion already happened
```
which avoids a lookup; overwriting the pre-existing key-value pair would change nothing, as a specific block hash can only correspond to a single block.

Reply:
I could also use the entry API here; it's not a blocker, though

Reply:
FWIW, the entry API solution would be:
```
pub fn insert(&mut self, key: K, value: V) -> bool {
    match self.map.entry(key.clone()) {
        Entry::Occupied(_) => return false,
        Entry::Vacant(entry) => {
            entry.insert(value);
        }
    }

    if let Some(Some(popped)) = self.queue.push(Some(key)) {
        self.map.remove(&popped);
    }

    true
}
```
however, I would prefer to introduce some benchmarks before potentially changing it.